From 58da2bbb5da1ed9091a38e60962bdb6149f55343 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Wed, 20 Sep 2023 17:51:18 +0200
Subject: [PATCH 0109/1010] x86: irq_pipeline: enable IPI shorthand for
 irq_send_oob_ipi()

Sending (oob) IPIs via apic_send_IPI_allbutself() is saner, enabling
the IPI shorthand when available from the platform in the same
move. Prefer this to calling the inner APIC handler directly.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 arch/x86/kernel/irq_pipeline.c |  7 -------
 arch/x86/kernel/smp.c          | 11 +++++++++++
 2 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/arch/x86/kernel/irq_pipeline.c b/arch/x86/kernel/irq_pipeline.c
index dc3b3eff15..73032a3be1 100644
--- a/arch/x86/kernel/irq_pipeline.c
+++ b/arch/x86/kernel/irq_pipeline.c
@@ -53,13 +53,6 @@ void handle_apic_irq(struct irq_desc *desc)
 	handle_oob_irq(desc);
 }
 
-void irq_send_oob_ipi(unsigned int ipi,
-		const struct cpumask *cpumask)
-{
-	apic->send_IPI_mask_allbutself(cpumask,	apicm_irq_vector(ipi));
-}
-EXPORT_SYMBOL_GPL(irq_send_oob_ipi);
-
 static irqentry_state_t pipeline_enter_rcu(void)
 {
 	irqentry_state_t state = {
diff --git a/arch/x86/kernel/smp.c b/arch/x86/kernel/smp.c
index 8e041f8bb9..3566d471d0 100644
--- a/arch/x86/kernel/smp.c
+++ b/arch/x86/kernel/smp.c
@@ -271,6 +271,17 @@ DEFINE_IDTENTRY_SYSVEC_PIPELINED(CALL_FUNCTION_SINGLE_VECTOR,
 	trace_call_function_single_exit(CALL_FUNCTION_SINGLE_VECTOR);
 }
 
+#ifdef CONFIG_IRQ_PIPELINE
+
+void irq_send_oob_ipi(unsigned int ipi,
+		const struct cpumask *cpumask)
+{
+	apic_send_IPI_allbutself(apicm_irq_vector(ipi));
+}
+EXPORT_SYMBOL_GPL(irq_send_oob_ipi);
+
+#endif
+
 static int __init nonmi_ipi_setup(char *str)
 {
 	smp_no_nmi_ipi = true;
-- 
2.30.2

